### ISSUE TITLE
[App Service] Update API is Broken

### ISSUE THREAD

**Issue by tombuildsstuff** [Azure/azure-rest-api-specs#1697]
:wave:

I've been trying to add support for [App Services to Terraform](https://github.com/terraform-providers/terraform-provider-azurerm/pull/344) - during which time I've encountered some bugs with the App Services/Web Apps API:

When using this request:

```
PUT /subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/tharvey-dev/providers/Microsoft.Web/sites/tharvey-appservice?api-version=2016-08-01&forceDnsRegistration=false&skipCustomDomainVerification=true&skipDnsRegistration=true&ttlInSeconds=60 HTTP/1.1
Host: management.azure.com
User-Agent: HashiCorp-Terraform-v0.10.0-dev
Content-Length: 249
Authorization: Bearer replaced-with-this-random-string---never-gonna-give-you-up---never-gonna-let-you-down
Content-Type: application/json
Accept-Encoding: gzip

{
	"location": "westeurope",
	"tags": {},
	"properties": {
		"enabled": true,
		"serverFarmId": "/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/tharvey-dev/providers/Microsoft.Web/serverfarms/tharvey-appserviceplan",
		"clientAffinityEnabled": true
	}
}
```

The following response is returned:

```
HTTP/1.1 400 Bad Request
Cache-Control: no-cache
Pragma: no-cache
Content-Length: 451
Content-Type: application/json; charset=utf-8
Expires: -1
Strict-Transport-Security: max-age=31536000; includeSubDomains
Server: Microsoft-IIS/8.0
X-AspNet-Version: 4.0.30319
X-Powered-By: ASP.NET
x-ms-ratelimit-remaining-subscription-writes: 1192
x-ms-request-id: 8a3a5d83-e2bc-401b-9ee2-5bcd1d0e7ab6
x-ms-correlation-request-id: 8a3a5d83-e2bc-401b-9ee2-5bcd1d0e7ab6
x-ms-routing-request-id: UKWEST:20170911T115952Z:8a3a5d83-e2bc-401b-9ee2-5bcd1d0e7ab6
Date: Mon, 11 Sep 2017 11:59:51 GMT

{
	"Code": "BadRequest",
	"Message": "ttlInSeconds object is not present in the request body.",
	"Target": null,
	"Details": [{
		"Message": "ttlInSeconds object is not present in the request body."
	}, {
		"Code": "BadRequest"
	}, {
		"ErrorEntity": {
			"ExtendedCode": "01008",
			"MessageTemplate": "{0} object is not present in the request body.",
			"Parameters": ["ttlInSeconds"],
			"Code": "BadRequest",
			"Message": "ttlInSeconds object is not present in the request body."
		}
	}],
	"Innererror": null
}
```

However - even when submitting the `ttlInSeconds` field in the request object (either, in the top level or within Parameters) - this response is returned. 

This is in contrast to initially Creating an App Service / Web App - when this argument is parsed from the URL (as is specified in the Swagger). I'm submitting this with the Go SDK - however this appears to be a Service Issue rather than a Swagger issue, so I'm filing it here.

Would it be possible for someone to take a look into this?

Thanks!

**Comment by olydis** [comment]
@Nking92 @naveedaz @EricSten-MSFT tagged everyone who seems to be involved with that API, feel free to assign more appropriately :-)

**Comment by missbruni** [comment]
any idea of when this is likely to be fixed?

**Comment by tombuildsstuff** [comment]
ping @Nking92 @naveedaz @EricSten-MSFT :)

**Comment by EricSten-MSFT** [comment]
@tombuildsstuff I only recently added the MSDeploy related APIs.  I believe @naveedaz or @davidebbo would know a better owner for this issue.

**Comment by naveedaz** [comment]
I’ll take a look today

**Comment by naveedaz** [comment]
From the request it seems you are trying to set ttlInSeconds? Setting ttl is a privileged operation for internal use. It probably should not be exposed in the swagger. 
The error you get is incorrect.  It talks about the value not specified in request body, when it should really it should be an UnAuthorized exception. I'll get create bugs on our side to track the swagger and error message change. 

What is your scenario for needing to specify the ttl ?

**Comment by tombuildsstuff** [comment]
@naveedaz thanks for taking a look at this.

We set these fields as they're present [in the CreateOrUpdate method signature in the Go SDK](https://raw.githubusercontent.com/Azure/azure-sdk-for-go/master/arm/web/apps.go) due to it's presence in the Swagger:

```
// CreateOrUpdate creates a new web, mobile, or API app in an existing resource group, or updates an existing app. This
// method may poll for completion. Polling can be canceled by passing the cancel channel argument. The channel will be
// used to cancel polling and any outstanding HTTP requests.
//
// resourceGroupName is name of the resource group to which the resource belongs. name is unique name of the app to
// create or update. To create or update a deployment slot, use the {slot} parameter. siteEnvelope is a JSON
// representation of the app properties. See example. skipDNSRegistration is if true web app hostname is not registered
// with DNS on creation. This parameter is
// only used for app creation. skipCustomDomainVerification is if true, custom (non *.azurewebsites.net) domains
// associated with web app are not verified. forceDNSRegistration is if true, web app hostname is force registered with
// DNS. TTLInSeconds is time to live in seconds for web app's default domain name.
func (client AppsClient) CreateOrUpdate(resourceGroupName string, name string, siteEnvelope Site, skipDNSRegistration *bool, skipCustomDomainVerification *bool, forceDNSRegistration *bool, TTLInSeconds string, cancel <-chan struct{}) (<-chan Site, <-chan error) {
```

> From the request it seems you are trying to set ttlInSeconds? Setting ttl is a privileged operation for internal use. It probably should not be exposed in the swagger.

Given these are currently accepted during the Create - are they defaulted to values internally if they're not set? Currently we submit some hard-coded values at Creation time (which are accepted) - so I'm wondering what the behaviour would be if they were removed (snippet from the Terraform codebase):

```
skipDNSRegistration := false
forceDNSRegistration := false
skipCustomDomainVerification := true
ttlInSeconds := "60"
_, createErr := client.CreateOrUpdate(resGroup, name, siteEnvelope, &skipDNSRegistration, &skipCustomDomainVerification, &forceDNSRegistration, ttlInSeconds, make(chan struct{}))
```

Thanks!

**Comment by tombuildsstuff** [comment]
ping @naveedaz - any update here? :)

**Comment by naveedaz** [comment]
Do not specify ttlInSeconds. Let me know if that works.

**Comment by tombuildsstuff** [comment]
@naveedaz unfortunately that's not an option since it's a required field in the Swagger

**Comment by tombuildsstuff** [comment]
ping @naveedaz - any update here? :)

**Comment by pixelicous** [comment]
@naveedaz @tombuildsstuff any update on this matter? stuck with terraform and arm templates 👎

**Comment by olydis** [comment]
@naveedaz any update? If `ttlInSeconds` is not supposed to be set by customers, it should **not** be documented in the Swagger, let alone be marked as required! This renders the operation completely useless.

**Comment by naveedaz** [comment]
This issue is address in api-version 2018-02-01. We will be releasing the swagger for this version by mid March.

**Comment by tombuildsstuff** [comment]
@naveedaz that's good to hear, thanks for confirming that. Are there any planned fixes to the existing API version? Unfortunately the Update API isn't usable with the Swagger in this state (as such, I believe a breaking swagger change would be acceptable here, given it never worked in the first place).

There's a separate question around deployment of the new API version - is there an estimated timeframe for the rollout to the Sovereign Clouds & Azure Stack in addition to Azure Public?

Thanks!

**Comment by olydis** [comment]
@naveedaz +1 on fixing the existing Swagger - *wrong* specifications must be fixed, changing something that is broken already is not really a breaking change

**Comment by pixelicous** [comment]
@naveedaz Could you plesae shade some light on the ETA for this? roadmap? this is breaking stuff on our end..

**Comment by naveedaz** [comment]
Here is the pull request for existing api-version: https://github.com/Azure/azure-rest-api-specs/pull/2664

**Comment by tombuildsstuff** [comment]
👋 hey @olydis @missbruni @naveedaz @pixelicous @EricSten-MSFT 

I can confirm that the Swagger changes introduced in #1429 have fixed this issue for us - we've just released v1.4.0 of the AzureRM Provider which includes support for updating both App Services and Function Apps; as such I'm going to close this issue.

Thanks!
